[upstream_utils] Use pathlib instead of os.path

Gold856 opened the pull request:
A noteworthy change is the replacement of the `dp.startswith(os.path.join(".", "subdir"))` pattern. pathlib doesn't offer something with similar semantics besides `match` and `full_match`, so there's now a helper function that replicates the behavior.

Other notable changes include the addition of type annotations to ensure code correctness, using == to check file names instead of `endswith` for clarity (`endswith` is still used to check extensions), manual walking and copying being refactored in googletest, memory, nanopb, protobuf, and sleipnir to use `walk_cwd_and_copy_if`, and matching functions being shortened to the point where they can just be inlined into the lambda.

calcmogul commented:
The cleanups look good. Just need to figure out why the script reruns are giving different results than before.

Gold856 commented:
Turns out there _was_ an important semantic difference when I replaced `dp.startswith(os.path.join(".", "subdir"))`. I've added a helper function into upstream_utils.py to retain the old behavior.

Gold856 commented:
Kinda debating how extensible I want `has_prefix` to be. I _could_ make it match on partial directory names as well, although I see low utility since you can just specify more directories. Naming suggestions are also welcome.

calcmogul commented on upstream_utils/upstream_utils.py:
```diff
@@ -135,7 +136,22 @@ def comment_out_invalid_includes(filename, include_roots):
+    Returns:
+    True if the path begins with the prefix, False otherwise.
+    """
+    return len(path.parts) >= len(prefix.parts) and True in [
+        path.parts[i] == prefix.parts[i] for i in range(len(prefix.parts))
+    ]
```
If you give `zip()` lists of length 3 and 2, you'll get tuples for the first 2 elements.
```suggestion
    return len(path.parts) >= len(prefix.parts) and all([
        p1 == p2 for p1, p2 in zip(path.parts, prefix.parts)
    ])
```

KangarooKoala commented on wpimath/src/main/native/thirdparty/eigen/include/Eigen/src/ThreadPool/CoreThreadPoolDevice.h:
Seems like an accident?

KangarooKoala reviewed (commented):
CI's waiting on merge conflicts with the Eigen update, so I can't tell if everything's fully fixed now.

calcmogul commented:
The ThreadPool folder should have been excluded, but the inclusion pattern was incorrect and included one of the ThreadPool files anyway.

auscompgeek commented on upstream_utils/protobuf.py:
```diff
@@ -2,235 +2,236 @@
-from upstream_utils import Lib, copy_to, walk_if
+from upstream_utils import Lib, walk_cwd_and_copy_if
 
 protobuf_lite_sources = set(
     [
```
Whilst we're here, these should've been set literals.

calcmogul commented on upstream_utils/llvm.py:
```diff
@@ -88,19 +88,19 @@ def run_global_replacements(wpiutil_llvm_files):
 
-def flattened_llvm_files(llvm, dirs_to_keep):
-    file_lookup = {}
+def flattened_llvm_files(llvm: Path, dirs_to_keep: list[str]):
```
```suggestion
def flattened_llvm_files(llvm: Path, dirs_to_keep: list[Path]):
```
